ovn/northd/ovn-northd.c

```diff
@@ -983,14 +985,36 @@ build_acls(struct ovn_datapath *od, struct hmap *lflows)
      * send all IP packets through the conntrack action, which handles
      * defragmentation, in order to match L4 headers. */
     if (has_stateful) {
-        /* Ingress and Egress Pre-ACL Table (Priority 100).
-         *
-         * Regardless of whether the ACL is "from-lport" or "to-lport",
-         * we need rules in both the ingress and egress table, because
-         * the return traffic needs to be followed. */
-        ovn_lflow_add(lflows, od, S_SWITCH_IN_PRE_ACL, 100, "ip", "ct_next;");
-        ovn_lflow_add(lflows, od, S_SWITCH_OUT_PRE_ACL, 100, "ip", "ct_next;");
-
+        HMAP_FOR_EACH (op, key_node, ports) {
+            if (op->od != od || !strcmp(op->nbs->type, "router")) {
+                /* Can't use ct() for router ports. Consider the following configuration: 
+                lp1(10.0.0.2) on hostA--ls1--lr0--ls2--lp2(10.0.1.2) on hostB, 
+                For a ping from lp1 to lp2, First, the response will go through ct() 
+                with a zone for lp2 in the ls2 ingress pipeline on hostB. 
+                That ct zone knows about this connection. Next, it goes through ct() 
+                with the zone for the router port in the egress pipeline of ls2 on hostB. 
+                This zone does not know about the connection, as the icmp request 
+                went through the logical router on hostA, not hostB. This would only work 
+                with distributed conntrack state across all chassis. */
+                continue;
+            }
+            ds_init(&matchIn);
+            ds_init(&matchOut);
+            ds_put_format(&matchIn, "ip && inport == %s", op->json_key);
+            ds_put_format(&matchOut, "ip && outport == %s", op->json_key);
+
+            /* Ingress and Egress Pre-ACL Table (Priority 100).
+             *
+             * Regardless of whether the ACL is "from-lport" or "to-lport",
+             * we need rules in both the ingress and egress table, because
+             * the return traffic needs to be followed. */
+            ovn_lflow_add(lflows, od, S_SWITCH_IN_PRE_ACL, 100, ds_cstr(&matchIn), "ct_next;");
+            ovn_lflow_add(lflows, od, S_SWITCH_OUT_PRE_ACL, 100, ds_cstr(&matchOut), "ct_next;");
+            
+            ds_destroy(&matchIn);
+            ds_destroy(&matchOut);
+        }
```

Comment:
Here's an alternative that would result in a lot less flows, I think.

You could still create the original flows at priority 100 that are removed in this patch.  Then, in the loop, create a new flow at a higher priority when the type *is* router, match on IP and the router port ID, and set the action to "next" instead of "ct_next".  

That way we only have new flows for each router port, which there should be less of.